File: analysis/folha_vs_servicos.py
from typing import Any

import pandas as pd
from sqlalchemy import text

from analysis import fontes_receita


def distribuicao_salarios(conn: Any, year: int) -> pd.DataFrame:
    """Retorna proventos positivos para exibição em histograma (estornos excluídos — apenas visualização)."""
    df = pd.read_sql_query(text("SELECT proventos FROM fct_pessoal WHERE ano = :ano"), conn, params={"ano": year})
    df["proventos"] = pd.to_numeric(df["proventos"].astype(str).str.replace(",", "."), errors="coerce")
    return df[df["proventos"] > 0].dropna()


def run(conn: Any, years: list[int], empresa_ids: list[str] | None = None) -> pd.DataFrame:
    # pessoal usa empresa='001' (código retornado pela API, não mapeado para as entidades do dashboard)
    # — a folha é um dado municipal consolidado e não deve ser filtrada por empresa_ids.
    empresa_clause = "AND empresa = ANY(:empresas)" if empresa_ids else ""
    records = []
    for year in years:
        folha = pd.read_sql_query(
            text("SELECT proventos FROM fct_pessoal WHERE ano = :ano"),
            conn,
            params={"ano": year},
        )
        total_folha = (
            pd.to_numeric(folha["proventos"].astype(str).str.replace(",", "."), errors="coerce").fillna(0).sum()
        )

        pago_params: dict = {"ano": year}
        if empresa_ids:
            pago_params["empresas"] = empresa_ids
        pago = pd.read_sql_query(
            text(f"SELECT pago FROM raw_porciuncula_prefeitura.despesas_por_orgao WHERE ano = :ano {empresa_clause}"),
            conn,
            params=pago_params,
        )
        total_pago = pd.to_numeric(pago["pago"].astype(str).str.replace(",", "."), errors="coerce").fillna(0).sum()

        rev_df = fontes_receita.run(conn, [year], empresa_ids=empresa_ids)
        rcl_proxy = float(rev_df.iloc[0]["total"]) if not rev_df.empty else 0.0

        percentual = total_folha / rcl_proxy * 100 if rcl_proxy > 0 else 0
        records.append(
            {
                "ano": year,
                "total_folha": total_folha,
                "total_pago": total_pago,
                "rcl_proxy": rcl_proxy,
                "percentual_folha": percentual,
            }
        )
    return pd.DataFrame(records)


def execucao_decimo_terceiro(conn: Any, year: int) -> dict[str, float] | None:
    """Calcula a execução orçamentária do 13º salário consultando despesas_gerais."""
    query = """
        SELECT
            SUM(empenhado) as empenhado_bruto,
            SUM(empenhado_liquido) as empenhado_liquido,
            SUM(liquidado) as liquidado,
            SUM(pago) as pago
        FROM fct_despesas
        WHERE ano = :ano
          AND elemento IN ('01', '03', '11', '96')
          AND (descricao ILIKE '%13%' OR descricao ILIKE '%decimo terceiro%' OR descricao ILIKE '%décimo terceiro%')
          AND descricao NOT ILIKE '%anula%'
          AND descricao NOT ILIKE '%136%'
          AND descricao NOT ILIKE '%137%'
          AND descricao NOT ILIKE '%138%'
          AND descricao NOT ILIKE '%139%'
          AND tipo_empenho != 'AN'
    """
    df = pd.read_sql_query(text(query), conn, params={"ano": year})
    if df.empty or pd.isna(df.iloc[0]["empenhado_bruto"]) or df.iloc[0]["empenhado_bruto"] is None:
        return None

    emp_bruto = float(df.iloc[0]["empenhado_bruto"])
    emp_liq = float(df.iloc[0]["empenhado_liquido"]) if not pd.isna(df.iloc[0]["empenhado_liquido"]) else emp_bruto
    liq = float(df.iloc[0]["liquidado"]) if not pd.isna(df.iloc[0]["liquidado"]) else 0.0
    pag = float(df.iloc[0]["pago"]) if not pd.isna(df.iloc[0]["pago"]) else 0.0

    pct_pago = (pag / emp_liq) if emp_liq > 0 else 0.0

    return {
        "empenhado": emp_liq,
        "empenhado_bruto": emp_bruto,
        "liquidado": liq,
        "pago": pag,
        "pct_pago": pct_pago,
    }


def detalhe_decimo_terceiro(conn: Any, year: int) -> pd.DataFrame:
    """Retorna o detalhamento da execução orçamentária do 13º salário por Órgão e Função."""
    query = """
        SELECT
            COALESCE(empresa_nome, empresa_id) as orgao,
            COALESCE(funcao_nome, 'Outros') as funcao,
            SUM(empenhado) as empenhado_bruto,
            SUM(empenhado_liquido) as empenhado_liquido,
            SUM(liquidado) as liquidado,
            SUM(pago) as pago
        FROM fct_despesas
        WHERE ano = :ano
          AND elemento IN ('01', '03', '11', '96')
          AND (descricao ILIKE '%13%' OR descricao ILIKE '%decimo terceiro%' OR descricao ILIKE '%décimo terceiro%')
          AND descricao NOT ILIKE '%anula%'
          AND descricao NOT ILIKE '%136%'
          AND descricao NOT ILIKE '%137%'
          AND descricao NOT ILIKE '%138%'
          AND descricao NOT ILIKE '%139%'
          AND tipo_empenho != 'AN'
        GROUP BY empresa_nome, empresa_id, funcao_nome
        ORDER BY pago DESC
    """
    df = pd.read_sql_query(text(query), conn, params={"ano": year})
    if df.empty:
        return pd.DataFrame(columns=["orgao", "funcao", "empenhado", "liquidado", "pago", "pct_pago"])

    df["empenhado_bruto"] = pd.to_numeric(df["empenhado_bruto"], errors="coerce").fillna(0.0)
    df["empenhado_liquido"] = pd.to_numeric(df["empenhado_liquido"], errors="coerce").fillna(0.0)
    df["liquidado"] = pd.to_numeric(df["liquidado"], errors="coerce").fillna(0.0)
    df["pago"] = pd.to_numeric(df["pago"], errors="coerce").fillna(0.0)
    df["pct_pago"] = (df["pago"] / df["empenhado_liquido"] * 100.0).where(df["empenhado_liquido"] > 0, 0.0)

    # Use empenhado_liquido as the main 'empenhado' column for UI
    df["empenhado"] = df["empenhado_liquido"]

    return df[["orgao", "funcao", "empenhado", "liquidado", "pago", "pct_pago"]]

File: packages/db/src/queries/folha_vs_servicos.ts
import { sql } from "kysely";
import { db } from "../client";
import { getFontesReceita } from "./fontes_receita";

export interface FolhaVsServicosRecord {
  ano: number;
  totalFolha: number;
  totalPago: number;
  rclProxy: number;
  percentualFolha: number;
}

export interface DecimoTerceiroExecucao {
  empenhado: number;
  empenhadoBruto: number;
  liquidado: number;
  pago: number;
  pctPago: number;
}

export async function getFolhaVsServicos(
  years: number[],
  empresaIds?: string[] | null,
): Promise<FolhaVsServicosRecord[]> {
  const records: FolhaVsServicosRecord[] = [];

  for (const year of years) {
    let totalFolha = 0;
    try {
      const resF =
        await sql`SELECT proventos FROM fct_pessoal WHERE ano = ${year}`.execute(
          db,
        );
      for (const r of (resF.rows as Record<string, unknown>[]) || []) {
        totalFolha +=
          parseFloat(String(r.proventos ?? "0").replace(",", ".")) || 0;
      }
    } catch {}

    let totalPago = 0;
    try {
      let qP = sql`SELECT pago FROM fct_despesas_por_orgao WHERE ano = ${year}`;
      if (empresaIds && empresaIds.length > 0) {
        qP = sql`${qP} AND empresa = ANY(${empresaIds})`;
      }
      const resP = await qP.execute(db);
      for (const r of (resP.rows as Record<string, unknown>[]) || []) {
        totalPago += parseFloat(String(r.pago ?? "0").replace(",", ".")) || 0;
      }
    } catch {}

    const revDf = await getFontesReceita([year], empresaIds);
    const rclProxy = revDf.length > 0 ? revDf[0].total : 0;
    const percentualFolha = rclProxy > 0 ? (totalFolha / rclProxy) * 100 : 0;

    records.push({
      ano: year,
      totalFolha,
      totalPago,
      rclProxy,
      percentualFolha,
    });
  }

  return records;
}

export async function getExecucaoDecimoTerceiro(
  year: number,
  empresaIds?: string[] | null,
): Promise<DecimoTerceiroExecucao | null> {
  try {
    let q = sql`
      SELECT
        SUM(CAST(REPLACE(empenhado, ',', '.') AS numeric)) as empenhado_bruto,
        SUM(CAST(REPLACE(empenhado_liquido, ',', '.') AS numeric)) as empenhado_liquido,
        SUM(CAST(REPLACE(liquidado, ',', '.') AS numeric)) as liquidado,
        SUM(CAST(REPLACE(pago, ',', '.') AS numeric)) as pago
      FROM fct_despesas
      WHERE ano = ${year}
        AND (
          descricao ILIKE '%13%' OR
          descricao ILIKE '%decimo terceiro%' OR
          descricao ILIKE '%décimo terceiro%' OR
          historico ILIKE '%13%' OR
          historico ILIKE '%decimo terceiro%' OR
          historico ILIKE '%décimo terceiro%'
        )
        AND descricao NOT ILIKE '%anula%'
        AND descricao NOT ILIKE '%136%'
        AND descricao NOT ILIKE '%137%'
        AND descricao NOT ILIKE '%138%'
        AND descricao NOT ILIKE '%139%'
        AND tipo_empenho != 'AN'
    `;
    if (empresaIds && empresaIds.length > 0) {
      q = sql`${q} AND empresa_id = ANY(${empresaIds})`;
    }
    const res = await q.execute(db);

    const r = (res.rows as Record<string, unknown>[])?.[0];
    if (!r || r.empenhado_bruto === null || r.empenhado_bruto === undefined)
      return null;

    const empBruto = parseFloat(String(r.empenhado_bruto ?? "0")) || 0;
    const empLiq = parseFloat(String(r.empenhado_liquido ?? "0")) || empBruto;
    const liq = parseFloat(String(r.liquidado ?? "0")) || 0;
    const pag = parseFloat(String(r.pago ?? "0")) || 0;
    const pctPago = empLiq > 0 ? pag / empLiq : 0;

    return {
      empenhado: empLiq,
      empenhadoBruto: empBruto,
      liquidado: liq,
      pago: pag,
      pctPago,
    };
  } catch {
    return null;
  }
}

export interface SalaryBin {
  faixa: string;
  min: number;
  max: number;
  count: number;
}

export async function getDistribuicaoProventos(
  year: number,
): Promise<SalaryBin[]> {
  try {
    const res = await sql`
      SELECT proventos
      FROM fct_pessoal
      WHERE ano = ${year}
    `.execute(db);

    const values: number[] = [];
    for (const r of (res.rows as Record<string, unknown>[]) || []) {
      const v = parseFloat(String(r.proventos ?? "0").replace(",", "."));
      if (!Number.isNaN(v) && v > 0) {
        values.push(v);
      }
    }

    const BINS = [
      { faixa: "R$ 0 - 2,5k", min: 0, max: 2500 },
      { faixa: "R$ 2,5k - 5k", min: 2500, max: 5000 },
      { faixa: "R$ 5k - 7,5k", min: 5000, max: 7500 },
      { faixa: "R$ 7,5k - 10k", min: 7500, max: 10000 },
      { faixa: "R$ 10k - 12,5k", min: 10000, max: 12500 },
      { faixa: "R$ 12,5k - 15k", min: 12500, max: 15000 },
      { faixa: "R$ 15k - 17,5k", min: 15000, max: 17500 },
      { faixa: "R$ 17,5k - 20k", min: 17500, max: 20000 },
      { faixa: "> R$ 20k", min: 20000, max: Infinity },
    ];

    const counts = BINS.map((b) => ({ ...b, count: 0 }));
    for (const val of values) {
      for (const b of counts) {
        if (val >= b.min && (val < b.max || b.max === Infinity)) {
          b.count += 1;
          break;
        }
      }
    }

    return counts;
  } catch {
    return [];
  }
}

export interface DepartmentalPayrollItem {
  descricao: string;
  pago: number;
}

export async function getDepartmentalPayroll(
  year: number,
  empresaIds?: string[] | null,
): Promise<DepartmentalPayrollItem[]> {
  try {
    let q = sql`
      SELECT
        descricao,
        SUM(CAST(REPLACE(pago, ',', '.') AS numeric)) as total_pago
      FROM fct_despesas_por_fornecedor
      WHERE ano = ${year}
        AND (descricao ~* ' E OUT(ROS?|\\.)' OR descricao ILIKE '%E OUTROS%' OR descricao ILIKE '%E OUTRO%')
    `;
    if (empresaIds && empresaIds.length > 0) {
      q = sql`${q} AND empresa = ANY(${empresaIds})`;
    }
    q = sql`${q} GROUP BY descricao ORDER BY total_pago DESC`;

    const res = await q.execute(db);
    return (res.rows as Record<string, unknown>[]).map((r) => ({
      descricao: String(r.descricao || ""),
      pago: parseFloat(String(r.total_pago ?? "0")) || 0,
    }));
  } catch {
    return [];
  }
}

export async function getPercentualChefiasEfetivas(
  year: number,
  _empresaIds?: string[] | null,
): Promise<number | null> {
  try {
    const res = await sql`
      SELECT
        COUNT(CASE WHEN (
          vinculo LIKE '%FG%' OR
          vinculo LIKE '%CC%' OR
          categoria_funcional = 'Efetivos ocupantes de cargo comissionado'
        ) THEN 1 END) AS efetivos_confianca,
        COUNT(CASE WHEN (
          categoria_funcional = 'Cargo comissionado extra-quadro' OR
          vinculo = 'Comissionado INSS' OR
          LOWER(vinculo) LIKE 'cargo comissionado%'
        ) THEN 1 END) AS comissionados_externos
      FROM fct_pessoal
      WHERE ano = ${year}
    `.execute(db);

    const row = (res.rows as Record<string, unknown>[])?.[0];
    const efetivosConfianca =
      parseFloat(String(row?.efetivos_confianca ?? "0")) || 0;
    const comissionadosExternos =
      parseFloat(String(row?.comissionados_externos ?? "0")) || 0;
    const totalConfianca = efetivosConfianca + comissionadosExternos;

    if (totalConfianca > 0) {
      return Number(((efetivosConfianca / totalConfianca) * 100).toFixed(1));
    }
  } catch {}
  return null;
}
